mountpoint-s3/src/data_cache/multilevel_cache.rs

```diff
@@ -0,0 +1,325 @@
+use std::sync::Arc;
+
+use async_trait::async_trait;
+use futures::task::{Spawn, SpawnExt};
+use tracing::{trace, warn};
+
+use crate::object::ObjectId;
+
+use super::{BlockIndex, ChecksummedBytes, DataCache, DataCacheResult};
+
+/// A data cache which uses both the local disk and S3 Express One Zone bucket as a storage.
+/// Disk cache is assumed to be faster so this is quiried first on `get_block` requests. An
+/// S3 Express One Zone cache is checked when data is missing on disk. Both caches are
+/// populated on `put_block`.
+pub struct MultilevelDataCache<DiskCache, ExpressCache, Runtime> {
+    disk_cache: Arc<DiskCache>,
+    express_cache: ExpressCache,
+    runtime: Runtime,
+}
+
+impl<DiskCache: DataCache, ExpressCache: DataCache, Runtime: Spawn>
+    MultilevelDataCache<DiskCache, ExpressCache, Runtime>
+{
+    pub fn new(disk_cache: Arc<DiskCache>, express_cache: ExpressCache, runtime: Runtime) -> Self {
+        // Method `MultilevelDataCache::block_size` relies on block sizes of both caches to be equal.
```

Comment:
A bit convoluted. We use the same blocks at both levels, so they need to have the same size. I'd mention it as a requirement in this method rustdoc.